### DISCUSSION TITLE
Strange behavior of channels when one channel used in a map function of another channel

### DISCUSSION THREAD

**Discussion by grsterin** [nextflow-io/nextflow#2027] [Q&A]
For some reason it looks like Nextflow doesn't wait for an output channel of a process to be populated with value when it is used in a `map` function called on another channel.

Consider this snippet:
```
nextflow.enable.dsl=2

workflow {

    results = produce_and_wait()

    results.view { "original channel: $it" }

    x = Channel.of(1).flatMap { results }

    x.view { "transformed channel: $it" }

}

process produce_and_wait {

    output:
    path ("test/file*")

    script:
    """
    mkdir test
    echo "123" > test/file1
    echo "345" > test/file2

    sleep 5
    """

}
```

Here's the output Nextflow produces when running it:
```
$ ./nextflow run ./qwerty.nf
N E X T F L O W  ~  version 20.10.0
Launching `./qwerty.nf` [happy_cuvier] - revision: 849a9eb95a
executor >  local (1)
[22/5ed855] process > produce_and_wait [100%] 1 of 1 ✔
transformed channel: DataflowVariable(value=null)
original channel: [/work/22/5ed855146835151eada4258db84501/test/file1, /work/22/5ed855146835151eada4258db84501/test/file2]
```

I expected that Nextflow would wait until `produce_and_wait` process is completed and only then it will print the result of the transformed channel. But in fact, first it prints the transformed channel (which has null value at this moment), then it completes the `produce_and_wait` process and after that it prints the original `results` channel.

Am I misunderstanding how channels work or is this is bug?

**Comment by pditommaso** [comment]
This makes no sense 

```
 x = Channel.of(1).flatMap { results }
```


I guess you wanted to do 

```
 x = results.flatMap()
```

**Reply by grsterin** [reply]
This was just a synthetic example mean to show that in this case `x.view()` is executed before the process which populates the `results` channel is completed.

In fact I wanted something like 
```
Channel
  .fromPath("/path/to/csv/file")
  .splitCsv(header: true)
  .map { row -> // do something with data from the results channel here based on row }
```
